src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1183,30 +1187,80 @@ static int32_t CopyFile_ReadWrite(int inFd, int outFd)
 }
 #endif // !HAVE_FCOPYFILE
 
-int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
+int32_t SystemNative_CopyFile(intptr_t sourceFd, const char* srcPath, const char* destPath, int32_t overwrite)
 {
     int inFd = ToFileDescriptor(sourceFd);
-    int outFd = ToFileDescriptor(destinationFd);
+    int outFd;
+    int ret;
+    struct stat_ sourceStat;
+
+    while ((ret = fstat_(inFd, &sourceStat)) < 0 && errno == EINTR);
+    if (ret != 0)
+    {
+        return -1;
+    }
+
+    struct stat_ destStat;
+    while ((ret = stat_(destPath, &destStat)) < 0 && errno == EINTR);
+    if (ret == 0)
+    {
+        if (sourceStat.st_dev == destStat.st_dev && sourceStat.st_ino == destStat.st_ino)
+        {
+            errno = EBUSY;
+            return -1;
+        }
+
+        if (!overwrite)
+        {
+            errno = EEXIST;
+            return -1;
+        }
+
+#if HAVE_CLONEFILE
+        // For clonefile we need to unlink the destination file first but we need to
+        // check permission first to ensure we don't try to unlink read-only file
+        if (access(destPath, W_OK) == 0)
+        {
+            ret = unlink(destPath);
+            if (ret != 0)
+            {
+                return ret;
+            }
+        }
+#endif
+    }
+
+#if HAVE_CLONEFILE
+    ret = clonefile(srcPath, destPath, 0);
+    if (ret == 0 || (errno != ENOTSUP && errno != EXDEV))
```

Comment:
If there's a race condition and the file is created after the above checks but before we call unlink, this could fail with EEXIST, right?  In which case shouldn't we fall back to the unoptimized case below?

I'm wondering if we should change this to just be:
```C
if (ret == 0)
{
    return ret;
}
// For all errors, fall through to attempt slow copy.
```

Reply:
It would likely make sense to allow `EEXIST` to fall through to avoid a race condition. We don't expect it to happen in any other case.

Most of the errors returned by `clonefile` are unrecoverable and would happen for the `open` later anyway. Not handling them here would open doors for more race conditions. They would likely be benign but it makes it harder to argue about the correctness.

The two errors that are handled there are for cases where `clonefile` doesn't work due to technical limitations (file system doesn't support cloning / copying between two different file systems or volumes).